src/eos-metrics-instrumentation.c

```diff
@@ -325,33 +395,47 @@ record_login (GDBusProxy *dbus_proxy,
               GVariant   *parameters,
               gpointer    user_data)
 {
-  if (strcmp ("SessionRemoved", signal_name) == 0)
-    {
-      const gchar *session_id;
-      g_variant_get (parameters, "(&s&o)", &session_id, NULL);
+  const gchar *session_id, *session_path;
+  guint32 user_id;
 
-      remove_session (session_id);
-    }
-  else if (strcmp ("SessionNew", signal_name) == 0)
+  if (strcmp ("SessionNew", signal_name) == 0)
     {
-      const gchar *session_id, *session_path;
       g_variant_get (parameters, "(&s&o)", &session_id, &session_path);
-
-      guint32 user_id;
       if (!get_user_id (session_path, &user_id))
         return;
 
       add_session (session_id, user_id);
     }
+  else if (strcmp ("SessionRemoved", signal_name) == 0)
+    {
+      /* SessionRemoved does not have session_path */
```

Comment:
Yes it does, that's the 'o' in the signal. https://www.man7.org/linux/man-pages/man5/org.freedesktop.login1.5.html documents:

```
               signals:
                 SessionNew(s session_id,
                            o object_path);
                 SessionRemoved(s session_id,
                                o object_path);
```

These signals contain the same information.

Reply:
Ah, but I see the problem is that when the session is removed, its object path is no longer valid, so the `get_user_id` method cannot be used to determine its user ID.

That's easy enough to deal with: maintain a second hash table whose keys are the session path strings, and values are the uint32 user ids.

Reply:
Another way to look at this: monitor the `UserNew` and `UserRemoved` signals. I think from reading https://www.freedesktop.org/wiki/Software/systemd/logind/ – but one should check! – that `UserNew` is fired when the first session is opened for a user, and `UserRemoved` is fired when they end their last session. Exactly what we want here.

Reply:
> Ah, but I see the problem is that when the session is removed, its object path is no longer valid, so the `get_user_id` method cannot be used to determine its user ID.
> 
> That's easy enough to deal with: maintain a second hash table whose keys are the session path strings, and values are the uint32 user ids.

Ya!  This is one way that I did at first time.  However, had some discussion in https://github.com/endlessm/eos-metrics-instrumentation/pull/97#discussion_r736821854

They both have pros & cons on each side.  I can do either way.  But, just cannot decide which one is better.

Reply:
> Another way to look at this: monitor the `UserNew` and `UserRemoved` signals. I think from reading https://www.freedesktop.org/wiki/Software/systemd/logind/ – but one should check! – that `UserNew` is fired when the first session is opened for a user, and `UserRemoved` is fired when they end their last session. Exactly what we want here.

It seems not that easy.

I place some debug messages on current commit:
```diff
$ git diff
diff --git a/src/eos-metrics-instrumentation.c b/src/eos-metrics-instrumentation.c
index 8ffe95a..c13b03b 100644
--- a/src/eos-metrics-instrumentation.c
+++ b/src/eos-metrics-instrumentation.c
@@ -317,6 +317,7 @@ add_session (const gchar *session_id,
   if (user_id < MIN_HUMAN_USER_ID)
     return;
 
+  g_warning ("User %u logs in with session ID %s", user_id, session_id);
   /*
    * The same user_id share the same UserSession here and only increase the
    * reference count of the UserSession, if the user already has an alive one.
@@ -367,6 +368,7 @@ remove_session (const gchar *session_id, enum RMSessionMethod rm)
   if (!session)
     return;
 
+  g_warning ("session ID %s logs out", session_id);
   if (!g_atomic_ref_count_dec (&session->rc))
     return;
 
```
According to the service's log:
```
Nov 05 11:57:52 endless systemd[1]: Started EndlessOS Metrics Instrumentation.
Nov 05 11:57:52 endless eos-metrics-ins[2905]: User 1000 logs in with session ID 2
Nov 05 11:58:47 endless eos-metrics-ins[2905]: User 1001 logs in with session ID 4
Nov 05 11:58:57 endless eos-metrics-ins[2905]: User 1001 logs in with session ID 6
Nov 05 11:59:02 endless eos-metrics-ins[2905]: session ID 6 logs out
Nov 05 11:59:03 endless eos-metrics-ins[2905]: session ID 4 logs out
Nov 05 11:59:33 endless eos-metrics-ins[2905]: session ID 2 logs out
Nov 05 11:59:33 endless systemd[1]: Stopping EndlessOS Metrics Instrumentation...
Nov 05 11:59:33 endless systemd[1]: eos-metrics-instrumentation.service: Succeeded.
Nov 05 11:59:33 endless systemd[1]: Stopped EndlessOS Metrics Instrumentation.
```
We can see the User 1001 logs in twice with different session ID 4 & 6.  Session ID 6 logs in & out during Session ID 4's living time.

Reply:
I agree that session 6 logs in and out during session 4's lifetime.

But if we switch from monitoring *sessions* to monitoring *logged-in users* we get the desired behaviour. I tested by monitoring all signals emitted by this object, then performing the following steps:

- Logging in over SSH as "ghostpiratelechuck"
- Logging in from GDM as "ghostpiratelechuck"
- Logging in from GDM as "shared"
- Logging out over SSH as "ghostpiratelechuck"
- Logging out from GNOME as "ghostpiratelechuck"
- Logging out from GNOME as "shared"

`ts` is a simple program which prefixes a timestamp to each line from stdin.

The full output is on https://gist.github.com/wjt/79a4004870ecec135c406478dce294dd because there are lots of uninteresting signals about sessions becoming idle (and perhaps in a perfect world we would not count time that the session is idle, but hey ho). But here are the interesting bits:

```
wjt@camille:~$ gdbus monitor --system --dest org.freedesktop.login1 --object-path /org/freedesktop/login1 | ts                      
Nov 05 11:27:42 Monitoring signals on object /org/freedesktop/login1 owned by org.freedesktop.login1
Nov 05 11:27:42 The name org.freedesktop.login1 is owned by :1.2

# SSH login (uid 1002 is ghostpiratelechuck)
Nov 05 11:28:05 /org/freedesktop/login1: org.freedesktop.login1.Manager.UserNew (uint32 1002, objectpath '/org/freedesktop/login1/user/_1002')
Nov 05 11:28:05 /org/freedesktop/login1: org.freedesktop.login1.Manager.SessionNew ('10', objectpath '/org/freedesktop/login1/session/_310')

# Switching to GDM login screen (uid 109 is Debian-gdm)
Nov 05 11:28:15 /org/freedesktop/login1: org.freedesktop.login1.Manager.UserNew (uint32 109, objectpath '/org/freedesktop/login1/user/_109')
Nov 05 11:28:15 /org/freedesktop/login1: org.freedesktop.login1.Manager.SessionNew ('c3', objectpath '/org/freedesktop/login1/session/c3')

# Second ghostpiratelechuck session
Nov 05 11:28:23 /org/freedesktop/login1: org.freedesktop.login1.Manager.SessionNew ('13', objectpath '/org/freedesktop/login1/session/_313')

# GDM session dies
Nov 05 11:28:29 /org/freedesktop/login1: org.freedesktop.login1.Manager.SessionRemoved ('c3', objectpath '/org/freedesktop/login1/session/c3')

# Switching back to GDM, new GDM session
Nov 05 11:28:35 /org/freedesktop/login1: org.freedesktop.login1.Manager.SessionNew ('c4', objectpath '/org/freedesktop/login1/session/c4')

# Logging in as UID 1000 (shared)
Nov 05 11:28:39 /org/freedesktop/login1: org.freedesktop.login1.Manager.UserNew (uint32 1000, objectpath '/org/freedesktop/login1/user/_1000')
Nov 05 11:28:39 /org/freedesktop/login1: org.freedesktop.login1.Manager.SessionNew ('14', objectpath '/org/freedesktop/login1/session/_314')

# GDM dies. I switch back to my main session so that I can terminate the SSH session
Nov 05 11:28:44 /org/freedesktop/login1: org.freedesktop.login1.Manager.SessionRemoved ('c4', objectpath '/org/freedesktop/login1/session/c4')
Nov 05 11:28:50 /org/freedesktop/login1: org.freedesktop.login1.Manager.SessionNew ('c5', objectpath '/org/freedesktop/login1/session/c5')
Nov 05 11:29:09 /org/freedesktop/login1: org.freedesktop.login1.Manager.SessionRemoved ('c5', objectpath '/org/freedesktop/login1/session/c5')
Nov 05 11:29:19 /org/freedesktop/login1: org.freedesktop.login1.Manager.UserRemoved (uint32 109, objectpath '/org/freedesktop/login1/user/_109')

# Terminating ghostpiratelechuck's SSH session
Nov 05 11:29:27 /org/freedesktop/login1: org.freedesktop.login1.Manager.SessionRemoved ('10', objectpath '/org/freedesktop/login1/session/_310')

# Back to GDM, switching to active ghostpiratelechuck session
Nov 05 11:29:37 /org/freedesktop/login1: org.freedesktop.login1.Manager.UserNew (uint32 109, objectpath '/org/freedesktop/login1/user/_109')
Nov 05 11:29:37 /org/freedesktop/login1: org.freedesktop.login1.Manager.SessionNew ('c6', objectpath '/org/freedesktop/login1/session/c6')
Nov 05 11:29:54 /org/freedesktop/login1: org.freedesktop.login1.Manager.SessionRemoved ('c6', objectpath '/org/freedesktop/login1/session/c6')

# Logged out of ghostpiratelechuck's last session
Nov 05 11:29:56 /org/freedesktop/login1: org.freedesktop.login1.Manager.SessionRemoved ('13', objectpath '/org/freedesktop/login1/session/_313')

# GDM respawns
Nov 05 11:29:56 /org/freedesktop/login1: org.freedesktop.login1.Manager.SessionNew ('c7', objectpath '/org/freedesktop/login1/session/c7')

# A few seconds later, uid 1002 (ghostpiratelechuck) gets UserRemoved
Nov 05 11:30:06 /org/freedesktop/login1: org.freedesktop.login1.Manager.UserRemoved (uint32 1002, objectpath '/org/freedesktop/login1/user/_1002')

# Switching back to shared account and closing that session
Nov 05 11:30:25 /org/freedesktop/login1: org.freedesktop.login1.Manager.SessionRemoved ('c7', objectpath '/org/freedesktop/login1/session/c7')
Nov 05 11:30:25 /org/freedesktop/login1: org.freedesktop.login1.Manager.SessionRemoved ('14', objectpath '/org/freedesktop/login1/session/_314')
Nov 05 11:30:26 /org/freedesktop/login1: org.freedesktop.login1.Manager.SessionNew ('c8', objectpath '/org/freedesktop/login1/session/c8')

# UserRemoved for shared (uid 1000)
Nov 05 11:30:36 /org/freedesktop/login1: org.freedesktop.login1.Manager.UserRemoved (uint32 1000, objectpath '/org/freedesktop/login1/user/_1000')
```

So. Rather than listening to `SessionNew` / `SessionRemoved` as your sample code continues to do, listen to `UserNew` and `UserRemoved`. Then no refcounting will be needed.

Reply:
Hmm ... `UserNew` and `UserRemoved` looks good!

Reply:
Have a new commit to catch the `UserNew` and `UserRemoved` signals.